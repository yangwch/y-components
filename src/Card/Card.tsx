import classNames from 'classnames';
import React, { CSSProperties, HTMLAttributes } from 'react';
import { settings } from '../utils/global';
import './style/index.less';

const cardPrefix = `${settings.prefix}-card`;

export interface CardProps extends Omit<HTMLAttributes<HTMLDivElement>, 'title'> {
  style?: CSSProperties;
  children?: React.ReactNode;
  title?: React.ReactNode;
  extra?: React.ReactNode;
  headStyle?: CSSProperties;
  bodyStyle?: CSSProperties;
  hoverable?: boolean;
}

function Card(props: CardProps) {
  const { children, title, extra, style, headStyle, bodyStyle, hoverable } = props;

  const clsNames = classNames(`${cardPrefix}`, {
    [`${cardPrefix}-hoverable`]: hoverable,
  });

  const head = <CardHead title={title} extra={extra} style={headStyle} />;
  const body = (
    <div className={`${cardPrefix}-body`} style={bodyStyle}>
      {children}
    </div>
  );
  return (
    <div className={clsNames} style={style}>
      {head}
      {body}
    </div>
  );
}

const CardHead = ({
  title,
  extra,
  style,
}: {
  title?: React.ReactNode;
  extra?: React.ReactNode;
  style?: CSSProperties;
}) => {
  if (!title && !extra) return null;
  return (
    <div style={style} className={`${cardPrefix}-head`}>
      <div className={`${cardPrefix}-head-title`}>{title}</div>
      <div className={`${cardPrefix}-head-extra`}>{extra}</div>
    </div>
  );
};

export default Card;
