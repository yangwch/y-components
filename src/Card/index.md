**Card**

```tsx
import React from 'react';
import { Card, Divider } from 'y-components';

const P = ({ children }) => <p style={{ padding: 5, margin: 0 }}>{children}</p>;
export default () => {
  return (
    <div>
      <Divider>Card Basic Usage & Hoverable</Divider>
      <Card title="Card Title" extra={<a href="#">More</a>} hoverable style={{ width: 300 }}>
        <P>Card Content</P>
        <P>Card Content</P>
        <P>Card Content</P>
        <P>Card Content</P>
      </Card>
    </div>
  );
};
```

**Without Title**

```tsx
import React from 'react';

import { Card, Divider } from 'y-components';

const P = ({ children }) => <p style={{ padding: 5, margin: 0 }}>{children}</p>;
export default () => {
  return (
    <div>
      <Divider>Default</Divider>
      <Card style={{ width: 300 }}>
        <P>Card Content</P>
        <P>Card Content</P>
        <P>Card Content</P>
        <P>Card Content</P>
      </Card>

      <Divider>Custom Body Style</Divider>
      <Card style={{ width: 300 }} bodyStyle={{ padding: 5, background: '#cccccc2a' }}>
        <P>Card Content</P>
        <P>Card Content</P>
        <P>Card Content</P>
        <P>Card Content</P>
      </Card>
      
    </div>
  );
};
```
