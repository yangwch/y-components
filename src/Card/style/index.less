@import '../../style/themes/index.less';

@card-cls: ~'@{prefix}-card';
@card-hoverable-hover-border: transparent;

.@{card-cls} {
  box-sizing: border-box;
  border: @card-border;
  border-radius: @card-border-radius;
  background: @card-background;

  &-head {
    box-sizing: border-box;
    padding: @card-head-padding;
    height: @card-head-height;
    border-bottom: @card-head-border;
    display: flex;
    &-title {
      flex: 1;
    }
  }

  &-body {
    padding: @card-padding-base;
  }

  &-hoverable {
    cursor: pointer;
    transition: box-shadow 0.3s, border-color 0.3s;

    &:hover {
      border-color: @card-hoverable-hover-border;
      box-shadow: @card-shadow;
    }
  }

  &-small {
    .@{card-cls}-head {
      padding: @card-head-padding-sm;
      height: @card-head-height-sm;
    }
    .@{card-cls}-body {
      padding: @card-padding-base-sm;
    }
  }
}