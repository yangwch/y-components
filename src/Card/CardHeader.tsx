import React, { CSSProperties } from 'react';
import { cardPrefix } from './Card';

const CardHeader = ({
  title,
  extra,
  style,
}: {
  title?: React.ReactNode;
  extra?: React.ReactNode;
  style?: CSSProperties;
}) => {
  if (!title && !extra) return null;
  return (
    <div style={style} className={`${cardPrefix}-head`}>
      <div className={`${cardPrefix}-head-title`}>{title}</div>
      <div className={`${cardPrefix}-head-extra`}>{extra}</div>
    </div>
  );
};

export default CardHeader;
