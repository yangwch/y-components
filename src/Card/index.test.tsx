import React from 'react';
import { render } from '@testing-library/react';
import { mountTest } from '../../test/shared/mountTest';

import Card from './index';

describe('Card', () => {
  mountTest(() => <Card />);
  mountTest(() => <Card>test</Card>);
  mountTest(() => <Card title="test"/>);
});

it('render correctly', () => {
  const testFn = () => render(<Card>Test</Card>);
  expect(testFn).not.toThrowError();

  const { container } = testFn();
  expect(container.firstChild).toMatchSnapshot();
});
