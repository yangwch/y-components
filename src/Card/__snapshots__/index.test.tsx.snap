// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<div
  class="yy-card"
>
  <div
    class="yy-card-body"
  >
    Test
  </div>
</div>
`;
